app/src/main/java/com/example/una/NavigationDrawerActivities/DonationHistoryActivity.java

```diff
@@ -63,39 +68,57 @@ public void onClick(View v) {
         // telling recyclerView to use linearLayout
         rvDonations.setLayoutManager(linearLayoutManager);
-        // giving an adapter to the recyclerView
         rvDonations.setAdapter(adapter);
-        // divides each item within in the recyclerView with a horizontal line
         DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(rvDonations.getContext(), linearLayoutManager.getOrientation());
         rvDonations.addItemDecoration(dividerItemDecoration);
 
-        // TODO -- endless scrolling here using endless recycler view holder
-
-        client = new FirestoreClient();
+        scrollListener = new EndlessRecyclerViewScrollListener(linearLayoutManager) {
+            @Override
+            public void onLoadMore(int page, int totalItemsCount, RecyclerView view) {
+                next.get()
+                        .addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
+                            @Override
+                            public void onSuccess(QuerySnapshot documentSnapshots) {
+                                loadMoreData(documentSnapshots);
+                            }
+                        });
+            }
+        };
+        rvDonations.addOnScrollListener(scrollListener);
         // FETCH DONATIONS FROM FIREBASE
         fetchDonations();
     }
 
+    private void loadMoreData(QuerySnapshot documentSnapshots) {
+        for (QueryDocumentSnapshot donationDoc : documentSnapshots) {
+            if(donationDoc.get("donor_id").toString().equals(client.getCurrentUser().getUid())) {
+                donations.add(new Donation(donationDoc.getData()));
+            }
+        }
+        adapter.notifyDataSetChanged();
+        // Get the last visible document
+        lastVisible = documentSnapshots.getDocuments()
+                .get(documentSnapshots.size() - 1);
+
+        next = client.getDonationsCollection()
+                .orderBy("time", Query.Direction.DESCENDING)
+                .startAfter(lastVisible)
+                .limit(itemsPerPageQuery);
+    }
+
     // fetch donations for user here
     private void fetchDonations() {
-        // get donations from current user from Firestore and create a new Donation object for each one
-        client.findDonationsByUserId(client.getCurrentUser().getUid(), new OnCompleteListener<QuerySnapshot>() {
-            @Override
-            public void onComplete(@NonNull Task<QuerySnapshot> task) {
-                if (task.isSuccessful()) {
-                    for (QueryDocumentSnapshot donationDoc : task.getResult()) {
-                        donations.add(new Donation(donationDoc.getData()));
-                        Log.d(TAG, donationDoc.getId() + " => " + donationDoc.getData());
-                        Log.i("DonationDoc ID: ", donationDoc.getId());
+        // Construct query for first 25 cities, ordered by population
+        Query first = client.getDonationsCollection()
```

Comment:
You are getting all donations on the app and then filtering to show the current user's donations. This means you won't be showing the 15 donations you requested. Also a security loop hole if you load more data than what the user should be able to see. Why not add the condition to only download donations for one user id